components/mailchimp/mailchimp.app.js

```diff
@@ -0,0 +1,369 @@
+const mailchimp = require("@mailchimp/mailchimp_marketing");
+const get = require("lodash/get");
+const retry = require("async-retry");
+
+module.exports = {
+  type: "app",
+  app: "mailchimp",
+  propDefinitions: {
+    timer: {
+      label: "Polling schedule",
+      description: "Pipedream polls Mailchimp for events on this schedule.",
+      type: "$.interface.timer",
+      default: {
+        //intervalSeconds: 15 * 60, // by default, run every 15 minutes.
+        intervalSeconds: 60
+      },
+    },
+  },
+  methods: {
+    _authToken() {
+      return this.$auth.oauth_access_token;
+    },
+    _server() {
+      return this.$auth.dc;
+    },
+    _initMailchimpClient() {
+      mailchimp.setConfig({
+        accessToken: this._authToken(),
+        server: this._server(),
+      });
+      return mailchimp;
+    },
+    async _withRetries(apiCall) {
+      const retryOpts = {
+        retries: 5,
+        factor: 2,
+      };
+      return retry(async (bail) => {
+        try {
+          return await apiCall();
+        } catch (err) {
+          const statusCode = get(err, [
+            "response",
+            "status",
+          ]);
+          if (!this._isRetriableStatusCode(statusCode)) {
+            bail(`
+              Unexpected error (status code: ${statusCode}):
+              ${JSON.stringify(err.response)}
+            `);
+          }
+          console.warn(`Temporary error: ${err.message}`);
+          throw err;
+        }
+      }, retryOpts);
+    },
+    _statusIsSent(status) {
+      return [
+        "status",
+      ].includes(status);
+    },
+    /**
+     * Create a webhook on the specified Audience List.
+     * @params {String} listId - The unique ID that identifies the Audience List you would like to
+     * create a webhook on.
+     * @params {String} webhookUrl - URL listening for the webhook event.
+     * @returns {_links: array, events: object, id: string, list_id: string, sources: object, url:
+     * string} Object with array of links for the webhook manipulation through the Mailchimp API,
+     * an object with flags of the Mailchimp Marketing indicates the events being watched for, an
+     * unique ID of the webhoom, the unique ID which identifies the Audience List where the webhoook
+     * was created, a `sources` object indicating whether the sources for the event are an admin
+     * user, an audience subscriber, or via the API, the URL registered for the webhook event.
+     */
+    async createWebhook(listId, webhookUrl, events, sources) {
+      const mailchimp = this._initMailchimpClient();
+      return await this._withRetries(() =>
+        mailchimp.lists.createListWebhook(listId, {
```

Comment:
Mailchimp responds with the error message:
```
"We couldn't verify the URL is working. Please double check and try again. HTTP Code: 400"
```

Reply:
hey @js07 , thanks for this.
i troubleshooted trying to find an issue with the code base from different angles.
in conclusion, i believe this is a problem with Pipedream's Sources new endpoints. Why with new sources? Because if I register a webhook on Mailchimp using the URL from an old endpoint, the webhook gets created correctly. If I use the endpoint that is returned by `$.interface.http` , registering the webook fails. For my testing, I hardcoded in the `http-base.js` file the URL `https://503b39783ae49ef8ae6a07b05f64100b.m.pipedream.net` which I copy pasted from one of my old sources (one I had used for Mailgun), and using this URL works.
That said, the issue might due to slow response to Mailchimp when they try to verify the URL. You can see that I hardcoded the URL in `http-base.js` and this registers properly with Mailchimp.

in details, this is what I did for my troubleshooting:
1. I created an empty source on a local file in my environment, test.js, to confirm I am properly importing attributes from `base.js` and `http-base.js`. in this empty source, I `console.out` variables returned by the $.interface.http, the $.auths variable, etc. just to make sure all things needed are in place
2. I arranged a workflow where I put all the methods used in the sources files, and creating the webhook using this workflow works. I tried adding a local webpage thorugh ngrok, and using the URL provided by ngrok to register, this worked. I also tried with a website of my own, and by the URL from old Sources mentioned before.
3. I added the instruction  `this.http.respond({      status: 200,    });` to acknoledge Mailchimp the eevent. I've seen other developers have used this in the `run` method of the source. This didn't help.

so, can you please have a look again putting special attention to the URL returned by `$.interface.http` and the time it takes to respond?

maybe this is helpful, i created a few workflows, one to return the Mailchimp audiences list, this is a prop for the `new-subscriber` source which I used for testing, a workflow to create a webhook, a workflow to delete a workflow, and a workflow to list workfows. Important, a URL must be used once in a webook list, otherwise Mailchimp will return 400 Bad request.

List audiences
https://pipedream.com/@sergio/mailchimp-list-audience-lists-p_q6CqoKg
Create webhook (based on the sources, http-based code)
https://pipedream.com/@sergio/mailchimp-create-webhook-test-new-subscribe-source-troubleshoot-copy-2-p_95C1LmQ
Delete webhook
https://pipedream.com/@sergio/mailchimp-delete-webhooks-legacy-code-p_3nCbvBL
List webhooks
https://pipedream.com/@sergio/mailchimp-list-webhooks-legacy-code-p_xMCvkGW

Reply:
When you added the instruction `this.http.respond({ status: 200, });` to the `run` method, did it also run in cases where `body` was empty. For example:
```js
  async run(event) {
    this.http.respond({
      status: 200,
    });
    const { body } = event;
    if (!body) {
      //Probably a Mailchimp response sent on webhook registration.
      console.log("No body present in event");
      return;
    }
    this.processEvent(body);
  },
```
```js
  async run(event) {
    const { body } = event;
    if (!body) {
      //Probably a Mailchimp response sent on webhook registration.
      this.http.respond({
        status: 200,
      });
      console.log("No body present in event");
      return;
    }
    this.processEvent(body);
  },
```
Just making sure because I think the conditional `!body` will be true when mailchimp is testing the webhook URL on registration.

Reply:
The following successfully deploys a Mailchimp source for me:
- Checkout https://github.com/PipedreamHQ/pipedream/pull/1289/commits/86ef7047bc822d0fd4a8da0c8daf91d256252d01
- Change the [`run` method in `common/http-based.js`](https://github.com/PipedreamHQ/pipedream/blob/86ef7047bc822d0fd4a8da0c8daf91d256252d01/components/mailchimp/sources/common/http-based.js#L86-L94) to
```js
  async run(event) {
    this.http.respond({
      status: 200,
    });
    const { body } = event;
    if (!body) {
      //Probably a Mailchimp response sent on webhook registration.
      console.log("No body present in event");
      return;
    }
    this.processEvent(body);
  },
```
- Deploy `new-subscriber.js` using the **Audience List Id** found here:
https://us20.admin.mailchimp.com/lists/members -> Settings -> Audience name and defaults
```
pd dev components/mailchimp/sources/new-subscriber/new-subscriber.js
```

Reply:
This has previously been addressed.